src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+                                constraintClauses: default,
+                                body: SyntaxFactory.Block(
+                                    SyntaxFactory.Token(
+                                        SyntaxFactory.TriviaList(),
+                                        SyntaxKind.OpenBraceToken,
+                                        SyntaxFactory.TriviaList(SyntaxFactory.EndOfLine(Environment.NewLine))),
+                                    SyntaxFactory.SingletonList(qe),
+                                    SyntaxFactory.Token(SyntaxKind.CloseBraceToken)),
+                                expressionBody: null);
+
+                            var localFunctionInvocation = SyntaxFactory.InvocationExpression(SyntaxFactory.IdentifierName(localFunctionName));
+                            SyntaxNode newParentExpressionStatement = parentStatement.ReplaceNode(source, localFunctionInvocation);
+                            documentUpdate = new DocumentUpdate(parentStatement, new[] { localFunctionDeclaration, newParentExpressionStatement });
+                            return true;
+                        }
+                    }
+                }
+
+                documentUpdate = default;
+                return false;
+            }
+
+            private bool TryGetTypeSyntax(ExpressionSyntax expression, out TypeSyntax typeSyntax)
+            {
+                var typeSymbol = _semanticModel.GetTypeInfo(expression, _cancellationToken).Type;
+                if (!ValidateTypeSymbol(typeSymbol))
+                {
+                    typeSyntax = default;
+                    return false;
+                }
+
+                typeSyntax = typeSymbol.GenerateTypeSyntax();
+                return true;
+            }
+
+            private static bool ValidateTypeSymbol(ITypeSymbol typeSymbol)
+            {
+                if (typeSymbol.IsAnonymousType || typeSymbol.TypeKind == TypeKind.Error)
+                {
+                    return false;
+                }
+
+                if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+                {
+                    foreach (var typeArgument in namedTypeSymbol.TypeArguments)
+                    {
+                        if (!ValidateTypeSymbol(typeArgument))
+                        {
+                            return false;
+                        }
+                    }
+                }
+
+                return true;
+            }
+
+            private static bool TryGetTypeSyntax(ITypeSymbol typeSymbol, out TypeSyntax typeSyntax)
+            {
+                if (!ValidateTypeSymbol(typeSymbol))
+                {
+                    typeSyntax = default;
+                    return false;
+                }
+
+                typeSyntax = typeSymbol.GenerateTypeSyntax();
+                return true;
+            }
+
+            private static bool TryFindParentAssignment(SyntaxNode node, out SyntaxNode result)
```

Comment:
It's called TryFindParentAssignemnt, but is really returning the containing statement.  name seems wrong.